src/main/groups.ts

```diff
@@ -0,0 +1,201 @@
+        const rowValues: any[] = cellValues[i];
+        const groupName: string = rowValues[GROUP_NAME_COLUMN_INDEX];
+        const groupMemberNames: string[] = getGroupMemberNames(rowValues);
+
+        if(groupName === "") {
+            // Skip empty entries
+            continue;
+        }
+
+        if(groupsMap.hasOwnProperty(groupName)) {
+            Logger.log(`Group ${groupName} has already been defined in the Groups table. Combining the two lists of members`);
+            groupsMap[groupName] = groupsMap[groupName].concat(groupMemberNames);
+        } else {
+            groupsMap[groupName] = groupMemberNames;
+        }
+    }
+
+    return groupsMap;
+}
+
+function getGroupMemberNames(rowValues: any[]): string[] {
+    const groupMemberNameList: string = rowValues[GROUP_MEMBER_NAMES_COLUMN_INDEX];
+    return groupMemberNameList.split(COMMA_DELIMITER).map((name) => name.trim()).filter((name) => name !== "");
+}
+
+function loadSupergroupsFromOnestop(loadedGroups: GroupsMap): GroupsMap {
+    const cellValues: any[][] = getCellValues(SUPERGROUPS_TAB_NAME);
+
+    let loadedSupergroups: GroupsMap = {};
+    const allSupergroups: SupergroupMap = parseSupergroups(cellValues);
+    const supergroupStack: string[] = Object.keys(allSupergroups);
+
+    // Loads alls Supergroups and their dependencies using DFS
+    while(supergroupStack.length > 0) {
+        const currentSupergroupName: string = supergroupStack.pop()!;
+        const currentSupergroup: Supergroup = allSupergroups[currentSupergroupName];
+
+        // Skips already loaded Supergroups
+        if(loadedSupergroups.hasOwnProperty(currentSupergroupName)) {
+            continue;
+        }
+
+        if(allSubgroupsHaveBeenLoaded(currentSupergroup, loadedGroups, loadedSupergroups, allSupergroups)) {
+            // Loads a Supergroup if all its dependent subgroups have already been loaded
+            const subgroupMembers: string[] = getAllMembersFromSubgroups(currentSupergroup.subgroups, loadedGroups, loadedSupergroups);
+            loadedSupergroups[currentSupergroup.name] = subgroupMembers;
+        } else {
+            // Push the current Supergoup as it needs to be reprocessed after all of its dependencies have been loaded
+            supergroupStack.push(currentSupergroupName);
+
+            // Push all dependencis that are Supergroups to the stack
+            const subgroups: string[] = currentSupergroup.subgroups;
+            for(const subgroup of subgroups) {
+                if(!isValidGroup(subgroup, allSupergroups, loadedGroups)) {
+                    // Skip invalid groups
+                    continue;
+                } else if(isSuperGroup(subgroup, allSupergroups) && !loadedSupergroups.hasOwnProperty(currentSupergroupName)) {
+                    // Add subgroups that are Supergroups that have not yet been loaded
+                    supergroupStack.push(subgroup);
+                }
+            }
```

Comment:
This section can be refactored so we don't have a nested loop within this function but it would require having the stack as an output parameter or have the refactored function return a list of subgroups to add to the stack and then use list concatenation to actually add them. Would appreciate some input here

Reply:
i think the refactor makes sense, a list concat isn't ugly so i don't see much of a readability cost, or am i missing something?

Reply:
Refactored with a list concat